ddl/db_test.go

```diff
@@ -2362,3 +2362,11 @@ func (s *testDBSuite) TestAddIndexForGeneratedColumn(c *C) {
 	s.mustExec(c, "alter table t add index idx_y(y1)")
 	s.mustExec(c, "alter table t drop index idx_y")
 }
+
+func (s *testDBSuite) TestIssue9100(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test_db")
+	tk.MustExec("create table employ (a int, b int) partition by range (b) (partition p0 values less than (1));")
+	_, err := tk.Exec("alter table employ add unique index  p_a (a);")
+	c.Assert(err.Error(), Equals, "[ddl:1503]A PRIMARY KEY must include all columns in the table's partitioning function")
```

Comment:
The error message should be `A UNIQUE INDEX must include all columns in the table's partitioning function`?

Reply:
And it seems that the error of `create table employ (a int, b int, unique index(a)) partition by range (b) (partition p0 values less than (1));` also has this problem.

Reply:
Can you add this tests ?
```sql
mysql> CREATE TABLE t1 (
    ->     col1 INT NOT NULL,
    ->     col2 DATE NOT NULL,
    ->     col3 INT NOT NULL,
    ->     col4 INT NOT NULL,
    ->     UNIQUE KEY (col1, col2)
    -> )
    ->
    -> PARTITION BY RANGE( col1 ) (
    ->   PARTITION p1 VALUES LESS THAN (11),
    ->   PARTITION p2 VALUES LESS THAN (15)
    -> );
Query OK, 0 rows affected (0.01 sec)

mysql> alter table t1 add unique index  p_col1 (col1);

mysql> CREATE TABLE t2 (
    ->     col1 INT NOT NULL,
    ->     col2 DATE NOT NULL,
    ->     col3 INT NOT NULL,
    ->     col4 INT NOT NULL,
    ->     UNIQUE KEY (col1, col3)
    -> )
    -> PARTITION BY RANGE( col1 + col3 ) (
    ->   PARTITION p1 VALUES LESS THAN (11),
    ->   PARTITION p2 VALUES LESS THAN (15)
    -> );
Query OK, 0 rows affected (0.05 sec)

mysql> alter table t2 add unique index  p_col1 (col1);
```

Reply:
This one I'll focus on alter table.
I'll fix create table in another PR, **it has a bug**, not simply the error message. @zimulala @ciscoxll

Reply:
Do we need to create an issue?

Reply:
https://github.com/pingcap/tidb/pull/9345